Circus + Celery - MVP

dannon opened the pull request:
This is *VERY* WIP; we wanted to get this open to ask questions about refactoring around the application context.   More documentation and description is forthcoming.

Right now it can do exactly one deferred task -- recalculating user disk usage.  But it works.  We're working on compiling a list of things to knock out and incorporate: https://github.com/galaxyproject/galaxy/issues/11721

Dependencies are *NOT* currently updated -- I couldn't get poetry to cooperate.  They are in pyproject.toml, but requirements are out of date, so every bit of this will fail CI (and your local testing, unless you `pip install circus celery uvicorn gunicorn` and maybe other stuff)

## What did you do? 
- A bare minimum circus + celery configuration.  VER

## Why did you make this change?
- https://github.com/galaxyproject/galaxy/issues/10893


## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [x] Instructions for manual testing are as follows:
  1. forthcoming

dannon committed:
Basic dev reloading w/ circus

dannon committed:
Initial circus addition -- dep update needs to be run still

dannon committed:
Starts a celery worker (that does nothing yet, but it does connect to my local redis)

dannon committed:
Split celery worker and tasks up

dannon committed:
Move recalculate task over.  Debug mode for celery, refactor a bit (still no bueno, just trying to get end-end complete)

dannon committed:
rebase all this out, but try to get galaxy's config loading correctly in this small context

dannon committed:
Swap to gunicorn w/ uvicorn worker

dannon committed:
Incremental config changes; this makes schema loader resolve correctly

dannon committed:
Sync up get_galaxy_app; seems the same functionally

dannon committed:
It actually works and does a recalc with this stripped down context -- we need a cleaner way to get this

dannon committed:
Trim down and clean up required context, add a log message to recalc task (it works!)

mvdbeek committed:
Hack in dynamic reading of broker from galaxy.yml

dannon commented on lib/galaxy/celery/tasks.py:
```diff
@@ -0,0 +1,40 @@
+    _preserved_argv = list(sys.argv[1:])
+    del sys.argv[1:]
+
+    registry = Registry()
```
@jmchilton I was hoping to ask you about this.  With the new DI stuff in mind, what's the right approach for acquiring the most minimal possible galaxy application context here (so, model, managers, config, security, and that's about it -- no threads)?

I looked at the fast_factory approach, but the whole app pair, etc., is way more than we need.  Marius and I were originally trying to directly use a UniverseApplication, but that also does more than we wanted, and I chopped it back to basically what I pulled out of db_shell to get it to work, but I'm positive this is not ideal.

jmchilton replied:
I think it is fine to just just create the database model mapping object - seems totally fine for a first pass. We do that in the tests and stuff. If you wanted a bit more I might try to to define an interface and extract UniverseApplication into a smaller superclass.

Here is a rough pass at that, I haven't run it because my laptop is currently dying... all the time so I'm sure it doesn't work at all really but just working through some ideas:

https://github.com/galaxyproject/galaxy/commit/3ba7ae24ead169e787dbd00138249d8261fa76d1

I think the biggest issue here is that each task is going to need a different sized app. In theory DI is perfect for that but it might be awkward to get to that point where dependencies are declared and injected as needed for each task. We need to have the components themselves not depend on ``app``, we need to figure out how to order shutdown tasks, etc..

Sorry I don't have a more complete answer given my endless stream of technical issues.

dannon replied:
@jmchilton No worries at all, thanks so much!  Marius took your tweak and ran with it and it's working great.

jdavcs committed:
Add circus; update all

mvdbeek committed:
Add celery

jmchilton committed:
WIP: minimal app?

mvdbeek committed:
Split UniverseApplication

mvdbeek committed:
Import MinimalGalaxyApplication

mvdbeek committed:
Revert "Split UniverseApplication"

This reverts commit 95f5ebd7122e0301c6458a11ba91068b4336734f.

mvdbeek committed:
Build minimal galaxy app, use in celery tasks

mvdbeek committed:
Use lru_cache to speed up getting app

mvdbeek committed:
Restructure so test collection should work

mvdbeek committed:
Use DI in celery task

mvdbeek committed:
Add enable_celery_tasks config option

mvdbeek committed:
Work around tool shed user controller importing celery task

mvdbeek committed:
Add purge dataset task

mvdbeek committed:
Break UniverseApplication into smaller parts

so we can use a variant of app that loads managers in Celery.

mvdbeek committed:
Let jop handler queue stop job

Unless jobs are not tracked in the database (a contender for deprecation
?).

mvdbeek committed:
Ensure connection pool is closed when forking

mvdbeek committed:
Don't override tempfile.tempdir when importing celery app

Fixes
```
Traceback (most recent call last):
  File "./scripts/functional_tests.py", line 127, in <module>
    driver_util.drive_test(find_test_driver())
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy_test/driver/driver_util.py", line 1151, in drive_test
    sys.exit(test_driver.run())
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy_test/driver/driver_util.py", line 940, in run
    self.setup()
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy_test/driver/driver_util.py", line 1007, in setup
    self._configure(config_object)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy_test/driver/driver_util.py", line 985, in _configure
    self.galaxy_test_tmp_dir = get_galaxy_test_tmp_dir()
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy_test/driver/driver_util.py", line 106, in get_galaxy_test_tmp_dir
    galaxy_test_tmp_dir = tempfile.mkdtemp()
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/tempfile.py", line 366, in mkdtemp
    _os.mkdir(file, 0o700)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpi834284l/tmp/tmpts6ajx8g'
```

mvdbeek committed:
Add celery to galaxy-app

nsoranzo commented on lib/galaxy/structured_app.py:
```diff
@@ -41,7 +41,43 @@ class BasicApp(Container):
+    host_security_agent: HostAgent
+
+
+class MininmalManagerApp(MinimalApp):
```
```suggestion
class MinimalManagerApp(MinimalApp):
```

dannon committed:
Update lib/galaxy/structured_app.py

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

dannon committed:
Use celery tasks directly

dannon committed:
Globally swap MininmalManagerApp to MinimalManagerApp (thanks for
noticing the first one @nsoranzo -- I am surprised how many of these
were in here)

mvdbeek commented on lib/galaxy/managers/hdas.py:
```diff
@@ -137,7 +137,7 @@ def copy_ldda(self, history, ldda, **kwargs):
     def purge(self, hda, flush=True):
         if self.app.config.enable_celery_tasks:
-            celery_app.send_task('galaxy.celery.tasks.purge_hda', kwargs={'hda_id': hda.id})
+            purge_hda.delay(hda_id=hda.id)
```
```
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/.venv/bin/celery", line 8, in <module>
    sys.exit(main())
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/__main__.py", line 15, in main
    sys.exit(_main())
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/bin/celery.py", line 213, in main
    return celery(auto_envvar_prefix="CELERY")
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/bin/base.py", line 132, in caller
    return f(ctx, *args, **kwargs)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/bin/worker.py", line 320, in worker
    worker = app.Worker(
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/worker/worker.py", line 94, in __init__
    self.app.loader.init_worker()
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/loaders/base.py", line 111, in init_worker
    self.import_default_modules()
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/loaders/base.py", line 106, in import_default_modules
    return [self.import_task_module(m) for m in self.default_modules]
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/loaders/base.py", line 106, in <listcomp>
    return [self.import_task_module(m) for m in self.default_modules]
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/loaders/base.py", line 86, in import_task_module
    return self.import_from_cwd(module)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/loaders/base.py", line 92, in import_from_cwd
    return import_from_cwd(
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/utils/imports.py", line 100, in import_from_cwd
    return imp(module, package=package)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.8/site-packages/celery/loaders/base.py", line 89, in import_module
    return importlib.import_module(module, package=package)
  File "/Users/mvandenb/.pyenv/versions/3.8.2/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/mvandenb/src/galaxy/lib/galaxy/celery/tasks.py", line 7, in <module>
    from galaxy.managers.hdas import HDAManager
  File "/Users/mvandenb/src/galaxy/lib/galaxy/managers/hdas.py", line 16, in <module>
    from galaxy.celery.tasks import purge_hda
ImportError: cannot import name 'purge_hda' from partially initialized module 'galaxy.celery.tasks' (most likely due to a circular import) (/Users/mvandenb/src/galaxy/lib/galaxy/celery/tasks.py)
```
I'll rebase this one out

dannon replied:
I am still working on it, please just hold on.

dannon committed:
Allow usage of old queue_worker recalc task on logout so this doesn't block when celery isn't enabled

dannon committed:
Reapply changes dropped in rebase shuffle

mvdbeek committed:
Do local import

jmchilton reviewed (approved):
That is a lovely shade of green! Very nice work.
